src/sgidClient.ts

```diff
@@ -0,0 +1,233 @@
+      scope: typeof scope === 'string' ? scope : scope.join(' '),
+      nonce: nonce ?? undefined,
+      state,
+      redirect_uri: redirectUri,
+      code_challenge: codeChallenge,
+      code_challenge_method: DEFAULT_SGID_CODE_CHALLENGE_METHOD,
+    })
+
+    const result: { url: string; nonce?: string } = { url }
+    if (nonce) {
+      result.nonce = nonce
+    }
+    return result
+  }
+
+  private getFirstRedirectUri(): string {
+    if (
+      !this.sgID.metadata.redirect_uris ||
+      this.sgID.metadata.redirect_uris.length === 0
+    ) {
+      throw new Error(Errors.MISSING_REDIRECT_URI_ERROR)
+    }
+    return this.sgID.metadata.redirect_uris[0]
+  }
+
+  /**
+   * Exchanges authorization code for access token.
+   * @param code The authorization code received from the authorization server
+   * @param nonce Nonce passed to authorizationUrl for this request. Specify null
+   * if no nonce was passed to authorizationUrl.
+   * @param redirectUri The redirect URI used in the authorization request. Defaults to the one
+   * passed to the SgidClient constructor.
+   * @param codeVerifier The code verifier that was used to generate the code challenge that was passed in `authorizationUrl`
+   * @returns The sub (subject identifier claim) of the user and access token. The subject
+   * identifier claim is the end-user's unique ID.
+   */
+  async callback({
+    code,
+    nonce = null,
+    redirectUri = this.getFirstRedirectUri(),
+    codeVerifier,
+  }: CallbackParams): Promise<{ sub: string; accessToken: string }> {
+    const tokenSet = await this.sgID.callback(
+      redirectUri,
+      { code },
+      { nonce: nonce ?? undefined, code_verifier: codeVerifier },
+    )
+    const { sub } = tokenSet.claims()
+    const { access_token: accessToken } = tokenSet
+    if (!sub) {
+      throw new Error(Errors.NO_SUB_ERROR)
+    }
+    if (!accessToken) {
+      throw new Error(Errors.NO_ACCESS_TOKEN_ERROR)
+    }
+    return { sub, accessToken }
+  }
+
+  /**
+   * Retrieves verified user info and decrypts it with your private key.
+   * @param accessToken The access token returned in the callback function
+   * @returns The sub of the end-user and the end-user's verified data
+   */
+  async userinfo({
+    accessToken,
+  }: UserInfoParams): Promise<{ sub: string; data: Record<string, string> }> {
+    /**
+     * sub: user sub (also returned previously in id_token)
+     * encryptedPayloadKey: key encrypted with client's public key (for decrypting userinfo jwe)
+     * data: jwe of userinfo
+     */
+    const {
+      sub,
+      key: encryptedPayloadKey,
+      data,
+    } = await this.sgID.userinfo<{
+      sub: string | undefined
+      key: string | undefined
+      data: Record<string, string> | undefined
+    }>(accessToken)
+
+    if (encryptedPayloadKey && data) {
+      const result = await this.decryptPayload(encryptedPayloadKey, data)
+      return { sub, data: result }
+    }
+
+    return { sub, data: {} }
```

Comment:
I know a while ago I commended jiehao for writing a good [PR description](https://github.com/opengovsg/sgid-client/pull/14) explaining why we return this when there's no key/data, but it would have been even better to leave a comment here to explain. shall we do that here?